### PR TITLE
Feature/block store logs

### PR DESCRIPTION
**Pull request by Aprogiena**
Adds detailed logging to classes related to block store. Unfortunately, many classes did not even have access to logging (logger factory / logger), so it was necessary to modify quite a lot of constructors. This then lead to change of files that called those constructors, so the final number of changed files is enormous. 

This is likely going to end with conflict hell (unless reviewed and merged quickly).

### CODE REVIEW

**Comment by Aprogiena** [comment]
@dangershony I've now found out that this PR is not ready for production as there is some kind of race condition:

- when I hit Ctrl+C to stop the node, after that the node crashed
- the reason for it was that 

```
Unhandled Exception: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LoggerFactory'.
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
```

This happened `at Stratis.Bitcoin.Connection.ConnectionManager..ctor`. So what happened here is that probably a new peer was discovered at the time when I hit Ctrl+C and the shutdown caused the `LoggerManager` to be disposed, but then `ConnectionManager` constructor was called with this disposed `LoggerManager` and thus the exception.

The full output is here:

```
info: Stratis.Bitcoin.FullNode[0]
      BlockStoreLoop.DownloadAndStoreBlocks stopping
^C
C:\StratisBitcoinFullNode\Stratis.BitcoinD>info: Stratis.Bitcoin.FullNode[0]
      MemoryPool.RelayWorker stopping
info: Stratis.Bitcoin.FullNode[0]
      BlockStore.RelayWorker stopping
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:69.84.42.56]:8333 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:138.201.152.245]:8333 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:122.56.90.87]:8333 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:73.202.111.95]:8333 offline
info: Stratis.Bitcoin.Base.BaseFeature[0]
      FlushAddressManager stopped
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:88.198.3.23]:8333 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:212.51.147.176]:8333 offline
info: Stratis.Bitcoin.Connection.ConnectionManager[0]
      Node [::ffff:52.26.194.229]:8333 offline
info: Stratis.Bitcoin.Base.BaseFeature[0]
      FlushChain stopped
info: Stratis.Bitcoin.Features.Consensus.ConsensusFeature[0]
      Flushing Cache CoinView...
info: Stratis.Bitcoin.Features.BlockStore.BlockStoreFeature[0]
      Flushing BlockStore...
info: Stratis.Bitcoin.Features.BlockStore.BlockStoreLoop[0]
      ====BlockStore Repository Stats(%)====
      Hit Count:      0
      Miss Count:     22611
      Delete Count:   0
      Insert Count:   10801
      Hit:            0,00 %
      Miss:           100,00 %
      =================================

      ====BlockStore Cache Stats(%)====
      Hit Count:      0
      Miss Count:     0
      Remove Count:   0
      Set Count:      0
      =================================


info: Stratis.Bitcoin.Features.MemoryPool.MempoolFeature[0]
      Saving Memory Pool...
info: Stratis.Bitcoin.Features.MemoryPool.MempoolFeature[0]
      ...Memory Pool Saved 0 transactions
fail: Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior[0]
      System.ObjectDisposedException: Cannot access a disposed object.
      Object name: 'LoggerFactory'.
         at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
         at Stratis.Bitcoin.Connection.ConnectionManager..ctor(Network network, NodeConnectionParameters parameters, NodeSettings nodeSettings, ILoggerFactory loggerFactory) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Connection\ConnectionManager.cs:line 72
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteExpressionBuilder.<>c__DisplayClass14_0.<Build>b__0(ServiceProvider provider)
         at Stratis.Bitcoin.Builder.FullNodeServiceProvider.<get_Features>d__2.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Builder\FullNodeServiceProvider.cs:line 38
         at System.Linq.Enumerable.<OfTypeIterator>d__33`1.MoveNext()
         at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
         at Stratis.Bitcoin.Base.ChainState.get_IsInitialBlockDownload() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Base\ChainState.cs:line 58
         at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<ProcessInvAsync>d__29.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Features\MemoryPool\MempoolBehavior.cs:line 322
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
         at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<AttachedNode_MessageReceived>d__26.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Features\MemoryPool\MempoolBehavior.cs:line 200

Unhandled Exception: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LoggerFactory'.
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
   at Stratis.Bitcoin.Connection.ConnectionManager..ctor(Network network, NodeConnectionParameters parameters, NodeSettings nodeSettings, ILoggerFactory loggerFactory) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Connection\ConnectionManager.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteExpressionBuilder.<>c__DisplayClass14_0.<Build>b__0(ServiceProvider provider)
   at Stratis.Bitcoin.Builder.FullNodeServiceProvider.<get_Features>d__2.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Builder\FullNodeServiceProvider.cs:line 38
   at System.Linq.Enumerable.<OfTypeIterator>d__33`1.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at Stratis.Bitcoin.Base.ChainState.get_IsInitialBlockDownload() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Base\ChainState.cs:line 58
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<ProcessInvAsync>d__29.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Features\MemoryPool\MempoolBehavior.cs:line 322
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<AttachedNode_MessageReceived>d__26.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Features\MemoryPool\MempoolBehavior.cs:line 216
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

```

Do you have any ideas how to fix this?

**Comment by Aprogiena** [comment]
I have an idea - `BaseFeature.Start` does this:
```
        public override void Start()
        {
            this.StartAddressManager();
            this.StartChain();

            var connectionParameters = this.connectionManager.Parameters;
            connectionParameters.IsRelay = this.nodeSettings.Mempool.RelayTxes;
            connectionParameters.TemplateBehaviors.Add(new ChainHeadersBehavior(this.chain, this.chainState));
            connectionParameters.TemplateBehaviors.Add(new AddressManagerBehavior(this.addressManager));

            this.disposableResources.Add(this.nodeSettings.LoggerFactory);
            this.disposableResources.Add(this.chainRepository);
            this.disposableResources.Add(this.connectionManager);
        }
```

maybe if we put `this.disposableResources.Add(this.nodeSettings.LoggerFactory);` as the last item there (after connectionManager), it would solve the problem, but I'm absolutely not sure about it and since I'm unable to reproduce the crash, I would need to confirm with you that it is a good idea and might actually solve this.

**Comment by Aprogiena** [comment]
Thinking about it, it is definitely a good idea to reverse that list of disposable resources, but it should not affect this problem - because CM constructor is called there, not using existing CM instance ...

So how to fix this is still unknown to me.

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs [hunk 1]
```diff
@@ -58,69 +63,96 @@ public BlockStoreBehavior(ConcurrentChain chain, IBlockRepository blockRepositor
...
+                        this.logger.LogTrace("(-)[CANCELED_EXCEPTION]");
+                        throw;
+                    }
 
 				// do nothing
 			}
 			catch (Exception ex)
 			{
-				this.logger.LogError(ex.ToString());
+				this.logger.LogError("Exception occurred: {0}", ex.ToString());
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: Perhaps we can remove the Debugger.Break();

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs [hunk 2]
```diff
@@ -131,66 +163,85 @@ private Task ProcessSendCmpctPayload(Node node, SendCmpctPayload sendCmpct)
...
 
-		////private Task ProcessGetBlocksAsync(Node node, GetBlocksPayload getBlocksPayload)
-		////{
-		////	ChainedBlock chainedBlock = this.chain.FindFork(getBlocksPayload.BlockLocators);
+            this.logger.LogTrace("(-)");
+        }
 
-		////	if (chainedBlock == null)
-		////		return Task.CompletedTask;
+        ////private Task ProcessGetBlocksAsync(Node node, GetBlocksPayload getBlocksPayload)
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: you can delete this commented method

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs [hunk 3]
```diff
@@ -275,34 +325,42 @@ public Task AnnounceBlocks(List<uint256> blockHashesToAnnounce)
...
-					var chainedBlock = this.chain.GetBlock(hashToAnnounce);
+					uint256 hashToAnnounce = blockHashesToAnnounce.Last();
+					ChainedBlock chainedBlock = this.chain.GetBlock(hashToAnnounce);
 					if (chainedBlock != null)
 					{
 						if (chainBehavior.PendingTip.GetAncestor(chainedBlock.Height) == null)
 						{
 							inventoryBlockToSend.Add(hashToAnnounce);
-						    this.logger.LogInformation($"sending inv peer={node.RemoteSocketEndpoint} hash={hashToAnnounce}");
+						    this.logger.LogDebug("Sending inventory hash '{0}' to peer '{1}'.", node.RemoteSocketEndpoint, hashToAnnounce);
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: Downgrading to LogDebug?

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 4]
```diff
@@ -229,24 +258,35 @@ internal void StartLoop()
                 var result = await this.stepChain.Execute(nextChainedBlock, disposeMode, cancellationToken);
                 if (result == StepResult.Stop)
                     break;
+
+                // TODO: Remove this as it does nothing?
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: Agree

**Comment by Aprogiena** [comment]
On of the solution that would fix the bug would be not to dispose logger factory at all.
However, this is not an elegant solution ...

The question is, can we find out a better one?

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs (hunk 3 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: Yes, I can't see why this should be INFO level

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs (hunk 2 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: k

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs (hunk 1 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreBehavior.cs]: k

**Comment by Aprogiena** [comment]
possible scenario of that crash:

1) i hit ctrl+c -> this initiated shutdown
2) this lead to `BaseFeature.Stop` which caused `LoggerFactory` to be disposed
3) however, there still is unstopped `MemoryPool` feature (i.e. MemoryPool should be stopped before BaseFeature but it is not) and that MemoryPool called `Bitcoin.Base.ChainState.get_IsInitialBlockDownload`
4) this contains line `var consensusFeature = this.fullNode.Services?.Features.OfType<IConsensusFeature>().FirstOrDefault(); // No hard failure if not present`
what this line does normally is that it selects all consensusFeatures that are registered
5) now somehow the DI mechanism sees that some of the Features (or Services, whatever) are disposed, therefore it thinks "hey, we don't have this, but we need it, let's create it"
6) and so it calls `ConnectionManager.ctor`
7) crash

**Pull request merged**